test/e2e/upgrade_management_cluster.go

```diff
@@ -0,0 +1,108 @@
+package e2e
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/config"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+)
+
+func upgradeManagementCluster() {
+	var (
+		specName = "clusterctl-upgrade"
+		managementClusterProxy framework.ClusterProxy
+	)
+	managementClusterProxy = bootstrapClusterProxy.GetWorkloadCluster(ctx, cluster.Namespace, cluster.Name)
```

Comment:
I'm a bit confused about this. This is how I think it is going:
1. The bootstrap cluster is created and initialized with the latest CAPM3 + BMO and Ironic
2. A workload cluster is created (completely empty)
3. In this upgrade test we say that the workload cluster is now a management cluster and do `clusterctl init` with an old version (but we never install BMO or Ironic here)
4. We upgrade CAPI and CAPM3 in this management/workload cluster and if it succeeds the test is passed.

Did I understand this correctly?

If so, I think we should rename this second management cluster so it is less confusing. The bootstrap cluster is also a management cluster so it is quite easy to confuse them. Maybe "upgradeCluster" or "testCluster" or something like this?

Second, I'm a bit worried that we are not installing Ironic and BMO and checking that things are working after the upgrade. Maybe we can add these later?

Reply:
Yes, it is correct and I agree and consider changing the cluster name. Also checking the cluster is working correctly would be in a follow up.